x/pool-incentives/keeper/keeper.go

```
@@ -320,40 +320,67 @@ func (k Keeper) GetAllGauges(ctx sdk.Context) []incentivestypes.Gauge {
+// If the gauge's DistributeTo.LockQueryType is ByDuration, it iterates over the lockableDurations, retrieves the pool ID for each duration, and checks if it matches the provided pool ID.
+// If none of the checks pass, it returns false, indicating that the pool is not incentivized.
+func (k Keeper) IsPoolIncentivized(ctx sdk.Context, providedPoolId uint64) bool {
+	epochDuration := k.incentivesKeeper.GetEpochInfo(ctx).Duration
+	lockableDurations := k.GetLockableDurations(ctx)
 
 	distrInfo := k.GetDistrInfo(ctx)
 
-	candidateGaugeIds := []uint64{}
-	for _, gaugeDuration := range lockableDurations {
-		gaugeId, err := k.GetPoolGaugeId(ctx, poolId, gaugeDuration)
-		if err == nil {
-			candidateGaugeIds = append(candidateGaugeIds, gaugeId)
-		}
-	}
-
 	for _, record := range distrInfo.Records {
-		for _, gaugeId := range candidateGaugeIds {
-			if record.GaugeId == gaugeId {
+		gauge, err := k.incentivesKeeper.GetGaugeByID(ctx, record.GaugeId)
+		if err != nil {
+			continue
+		}
+		if gauge.DistributeTo.LockQueryType == lockuptypes.ByGroup {
+			group, err := k.incentivesKeeper.GetGroupByGaugeID(ctx, record.GaugeId)
+			if err != nil {
+				continue
+			}
+			groupGauge, err := k.incentivesKeeper.GetGaugeByID(ctx, group.GroupGaugeId)
+			if err != nil {
+				continue
+			}
+			if !groupGauge.IsPerpetual {
+				// if the group is not perpetual, it is an externally incentivized gauge so we skip it
+				continue
+			}
+			poolIds, _, err := k.incentivesKeeper.GetPoolIdsAndDurationsFromGaugeRecords(ctx, group.InternalGaugeInfo.GaugeRecords)
+			if err != nil {
+				continue
+			}
+			for _, poolId := range poolIds {
+				if poolId == providedPoolId {
+					return true
+				}
+			}
+		} else if gauge.DistributeTo.LockQueryType == lockuptypes.NoLock {
+			poolId, err := k.GetPoolIdFromGaugeId(ctx, record.GaugeId, epochDuration)
+			if err != nil {
+				continue
```

Can we bubble up the error? I don't understand why we would be silently skipping in all these cases

---

A continuation of old logic, as described here https://github.com/osmosis-labs/osmosis/issues/6617

---

I would like to suggest attempting to avoid skipping errors in this PR because I see no reason for us to skip them under normal conditions.

If we discover a reason for why this is blocking, let's revisit

I'm afraid that if we leave this and keep punting the created issue, this tech debt that we cannot explain will only accumulate, further causing confusion in the future

---

The following two PRs should both address this comment and address #6617 

https://github.com/osmosis-labs/osmosis/pull/6604/commits/b5dbd6ce2fdcb85b487aa818e4d987249ab58afc
https://github.com/osmosis-labs/osmosis/pull/6604/commits/4b0428c87b08762e8cb15b26a17502e31db0662e

---

Ah! I finally realize why the previous code was skipping the error silently. It was doing some really really silly. For EVERY gauge ID, it was looping through EVERY lockable duration. If you think about it, a single gauge ID can only be linked to a single duration! So when checking a byDuration gauge, it would error on every loop except for the the duration which the gauge actually was.

What I do instead now is, pull the gauge's duration directly from the gauge itself, do a sanity check that the duration is one of the lockable durations, and then continue from there. Should result in significantly less iterations. @p0mvn